cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/relational/SQLDialectConversionExpressionFactory.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.datapipeline.relational;
+
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.cdap.etl.api.engine.sql.StandardSQLCapabilities;
+import io.cdap.cdap.etl.api.relational.Capability;
+import io.cdap.cdap.etl.api.relational.Expression;
+import io.cdap.cdap.etl.api.relational.ExpressionFactory;
+import io.cdap.cdap.etl.api.relational.ExpressionFactoryType;
+import io.cdap.cdap.etl.api.relational.InvalidExtractableExpression;
+import io.cdap.cdap.etl.api.relational.StringExpression;
+import io.cdap.cdap.etl.api.relational.StringExpressionFactoryType;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * An expression factory which converts a SQL {@link Expression} from one SQL dialect to another on compiling it.
+ */
+public class SQLDialectConversionExpressionFactory implements ExpressionFactory<String> {
+
+    SQLDialect sourceDialect;
+    SQLDialect destinationDialect;
+    @Nullable Schema schema;
+    SQLDialectConverter converter;
+
+    private static final Set<Capability> CAPABILITIES = Collections.unmodifiableSet(
+            new HashSet<Capability>() {{
+                add(StringExpressionFactoryType.SQL);
+                add(StandardSQLCapabilities.BIGQUERY);
+                add(StandardSQLCapabilities.SPARK);
```

Comment:
Do you expect this factory to understand BQ and Spark? I would say this should have capabilities according to source dialect

Reply:
Changed to dynamically add capabilities based on dialects.